provider/openstack/firewaller.go

```diff
@@ -172,32 +173,18 @@ func instServerID(inst instances.Instance) (string, error) {
 	return openstackName[lastDashPos+1:], nil
 }
 
-func deleteSecurityGroupsMatchingName(
+func (c *neutronFirewaller) deleteSecurityGroupsMatchingName(
 	ctx context.ProviderCallContext,
-	deleteSecurityGroups func(ctx context.ProviderCallContext, match func(name string) bool) error,
-	prefix string,
+	deleteSecurityGroups func(ctx context.ProviderCallContext, groups []neutron.SecurityGroupV2, match func(name string) bool) error,
+	match func(name string) bool,
 ) error {
-	re, err := regexp.Compile("^" + prefix)
+	neutronClient := c.environ.neutron()
+	securityGroups, err := neutronClient.ListSecurityGroupsV2()
 	if err != nil {
-		return errors.Trace(err)
-	}
-	return deleteSecurityGroups(ctx, re.MatchString)
-}
-
-func deleteSecurityGroupsOneOfNames(
-	ctx context.ProviderCallContext,
-	deleteSecurityGroups func(ctx context.ProviderCallContext, match func(name string) bool) error,
-	names ...string,
-) error {
-	match := func(check string) bool {
-		for _, name := range names {
-			if check == name {
-				return true
-			}
-		}
-		return false
+		handleCredentialError(err, ctx)
+		return errors.Annotate(err, "cannot list security groups")
 	}
-	return deleteSecurityGroups(ctx, match)
+	return deleteSecurityGroups(ctx, securityGroups, match)
```

Comment:
Before calling this, filter the list of security groups. We then don't need to pass along the match closure.